### ISSUE TITLE
Refactor Power Allocation Optimization: Replace CVXOPT with CVXPY + CuCLARABEL

### ISSUE THREAD

**Issue by thienandangthanh** [thienandangthanh/ASOPA#12]
## Summary

Replace CVXOPT-based power allocation optimization with CVXPY + CuCLARABEL backend to leverage GPU acceleration and improve computational performance.

### Current Bottleneck

The ASOPA training pipeline calls power allocation optimization thousands of times per epoch. Each call to `get_optimal_p()` in `resource_allocation_optimization.py` uses CVXOPT's CPU-based convex programming solver, which becomes a significant computational bottleneck during training.

**Performance Impact:**
- Power allocation is called once per training sample during cost computation (`problem_noop.py:62`)
- With batch size 64 and 1000+ batches per epoch, this results in 64,000+ optimization calls
- CPU-only optimization limits training throughput and GPU utilization

### Proposed Solution

Migrate to **CVXPY with CuCLARABEL backend** to:
1. **GPU acceleration**: Leverage NVIDIA CUDA for parallel optimization
2. **Better API**: More maintainable and readable problem formulation
3. **Automatic reformulation**: CVXPY handles conic form conversion automatically
4. **Future-proof**: Easier to switch solvers and experiment with alternatives

### Current CVXOPT Usage

The code solves two types of convex optimization problems:

#### 1. α = 1 (Logarithmic Fairness)
**Location**: `_get_optimal_p_alpha_1()` (lines 561-702)

**Problem Structure**:
```
minimize   -Σᵢ xᵢ
subject to log(noise/gᵢe^yᵢ) + log(2^(e^(xᵢ/wᵢ)) - 1) ≤ 0  (NOMA SIC constraints)
           e^yᵢ ≤ pᵢᵐᵃˣ                                   (power limits)
           xᵢ ≤ 10                                         (throughput upper bound)
```

**Solver Used**: `solvers.cp()` with custom gradient/Hessian

#### 2. α ∈ [0,1) (Proportional Fairness)
**Location**: `_get_optimal_p_alpha_0_1()` (lines 711-804)

**Problem Structure**:
```
maximize   Σᵢ wᵢ * (log₂(1 + gᵢpᵢ/Iᵢ))^(1-α) / (1-α)
subject to 0 ≤ pᵢ ≤ pᵢᵐᵃˣ                              (power limits)
           rᵢ ≥ 0                                       (throughput non-negative)
           -log₂(Iᵢ + gᵢpᵢ) ≥ rᵢ - c                    (linearized constraints)
```

**Solver Used**: `solvers.cp()` with Successive Convex Approximation (SCA)

### Why SDP Support Is Not Required

**Clarification**: Despite CuCLARABEL not supporting Semidefinite Programming (SDP), this refactoring is **feasible** because:

1. ✅ Both optimization problems are **convex programs** (not SDPs)
2. ✅ Constraints involve exponential/logarithmic functions (can be reformulated using exponential cones)
3. ✅ No semidefinite constraints (PSD matrices) are present in the formulation
4. ✅ CuCLARABEL supports: LP, QP, SOCP, and **exponential cone programming** - sufficient for our needs

### Phase 1: Setup and Dependencies

**Tasks:**
1. Add CVXPY and CuCLARABEL to `requirements.txt`
2. Verify GPU environment setup (CUDA 12.1+ compatibility)
3. Create feature branch: `feature/cvxpy-cuclarabel-optimization`

**Dependencies to add:**
```txt
cvxpy>=1.4.0
clarabel>=0.6.0        # CPU backend
cuclarabel>=0.1.0      # GPU backend (requires CUDA)
```

#### 2.1: Refactor α ∈ [0,1) Case (Simpler)

**Create**: `resource_allocation_optimization_cvxpy.py`

**Key Changes**:
```python
import cvxpy as cp
import clarabel

def _get_optimal_p_alpha_0_1_cvxpy(users, alpha, noise):
    """CVXPY reformulation with exponential cone constraints."""
    n = len(users)
    
    # Decision variables
    p = cp.Variable(n, pos=True)  # Power allocation
    r = cp.Variable(n, pos=True)  # Throughput auxiliary variables
    
    # Objective: maximize weighted alpha-fair throughput
    objective = cp.Maximize(
        cp.sum([users[i].w * cp.power(r[i], 1-alpha) / (1-alpha) 
                for i in range(n)])
    )
    
    # Constraints
    constraints = []
    for i in range(n):
        # Power limits
        constraints.append(p[i] <= users[i].p_max)
        
        # NOMA throughput constraints (using exponential cone)
        # log₂(1 + gᵢpᵢ/Iᵢ) ≥ rᵢ
        interference = noise + cp.sum([users[j].g * p[j] for j in range(i+1, n)])
        constraints.append(
            cp.constraints.exponential.ExpCone(
                interference, 
                users[i].g * p[i], 
                interference * cp.power(2, r[i])
            )
        )
    
    # Solve with CuCLARABEL
    problem = cp.Problem(objective, constraints)
    problem.solve(solver=cp.CLARABEL, verbose=False, gpu=True)
    
    return p.value
```

**Note**: The actual reformulation requires careful handling of SCA iterations and exponential cone formulations. This is a conceptual sketch.

#### 2.2: Refactor α = 1 Case (More Complex)

**Challenges**:
- Nested exponentials: `2^(e^(x/w))`
- Requires logarithm of log transformations
- May need additional auxiliary variables

**Approach**:
1. Introduce auxiliary variables for intermediate exponentials
2. Use exponential cone constraints: `(x, y, z) ∈ K_exp ⟺ y*e^(x/y) ≤ z`
3. Chain multiple cones for nested transformations

#### 3.1: Backward Compatibility

**Strategy**: Add solver selection parameter
```python
def get_optimal_p(users, alpha, noise, solver='cvxopt'):
    """
    Get optimal power allocation.
    
    Args:
        solver: 'cvxopt' (default, CPU) or 'cuclarabel' (GPU-accelerated)
    """
    if solver == 'cuclarabel':
        return _get_optimal_p_cvxpy(users, alpha, noise)
    else:
        # Existing CVXOPT implementation
        return _get_optimal_p_cvxopt(users, alpha, noise)
```

#### 3.2: Testing Strategy

**Unit Tests** (`tests/test_power_allocation.py`):
```python
def test_cvxpy_cvxopt_parity():
    """Verify CVXPY and CVXOPT produce equivalent results."""
    users = generate_topology(user_number=5)
    alpha = 0.5
    noise = 3.981e-15
    
    p_cvxopt = get_optimal_p(users, alpha, noise, solver='cvxopt')
    p_cvxpy = get_optimal_p(users, alpha, noise, solver='cuclarabel')
    
    # Check power allocations match within tolerance
    np.testing.assert_allclose(p_cvxopt, p_cvxpy, rtol=1e-3)
    
    # Check objective values match
    obj_cvxopt = get_objective_throughput(users, p_cvxopt, alpha, noise)
    obj_cvxpy = get_objective_throughput(users, p_cvxpy, alpha, noise)
    assert abs(obj_cvxopt - obj_cvxpy) < 1e-4

def test_gpu_acceleration():
    """Benchmark CPU vs GPU performance."""
    users = generate_topology(user_number=10)
    
    start = time.time()
    for _ in range(100):
        get_optimal_p(users, 0.5, 3.981e-15, solver='cvxopt')
    cpu_time = time.time() - start
    
    start = time.time()
    for _ in range(100):
        get_optimal_p(users, 0.5, 3.981e-15, solver='cuclarabel')
    gpu_time = time.time() - start
    
    print(f"CPU: {cpu_time:.2f}s, GPU: {gpu_time:.2f}s")
    print(f"Speedup: {cpu_time/gpu_time:.2f}x")
```

**Integration Tests**:
1. Run full training loop with both solvers
2. Verify convergence behavior matches
3. Check final model performance equivalence

**Validation Tests**:
1. Test on validation dataset (`ASOPA_validation.py`)
2. Compare against baseline methods
3. Ensure numerical stability across problem sizes (n=5 to n=20)

#### 4.1: Warm-Starting

**Optimization**: Use previous solution as warm-start for next iteration
```python
# In SCA loop
problem.solve(solver=cp.CLARABEL, warm_start=True, gpu=True)
```

#### 4.2: Batch Optimization

**Consideration**: Process multiple problem instances in parallel
```python
def get_optimal_p_batch(users_batch, alpha, noise):
    """Solve multiple power allocation problems in parallel on GPU."""
    # Reformulate as single large problem with block structure
    # Leverage GPU parallelism across samples
    pass
```

**Note**: Requires careful memory management and may need CuCLARABEL API extensions.

#### 4.3: Problem Size Adaptation

**Strategy**: Use CPU solver for small problems (n < 8), GPU for larger
```python
def get_optimal_p(users, alpha, noise, solver='auto'):
    if solver == 'auto':
        solver = 'cuclarabel' if len(users) >= 8 else 'cvxopt'
    # ...
```

### Phase 5: Documentation and Migration

**Documentation Updates**:
1. Update `docs/ARCHITECTURE.md`: Document new solver architecture
2. Update `docs/API_REFERENCE.md`: Add solver parameter documentation
3. Update `AGENTS.md`: Add guidance for solver selection
4. Create `docs/OPTIMIZATION_BACKEND.md`: Detailed technical documentation

**Configuration Updates**:
```python
# conf.py
parser.add_argument('--solver', type=str, default='cvxopt', 
                    choices=['cvxopt', 'cuclarabel', 'auto'],
                    help='Optimization solver backend')
parser.add_argument('--solver_gpu', action='store_true',
                    help='Enable GPU acceleration for solver')
```

**Migration Path**:
1. Keep CVXOPT as default in initial release
2. Add `--solver cuclarabel` flag for opt-in testing
3. After validation period, switch default to `'auto'`
4. Deprecate CVXOPT in future release (6+ months)

### Performance Improvements

**Conservative Estimates** (n=10 users):
- **Per-solve speedup**: 2-5x (GPU vs CPU)
- **End-to-end training**: 1.5-3x faster (optimization is ~50-70% of training time)
- **Validation time**: 2-4x faster

**Scaling Benefits** (n=15-20 users):
- Larger problems benefit more from GPU parallelism
- Expected 5-10x speedup for n > 15

### Code Quality Improvements

1. **Readability**: CVXPY's domain-specific language is more intuitive
2. **Maintainability**: Less manual gradient/Hessian computation
3. **Debugging**: Better error messages and constraint violation diagnostics
4. **Flexibility**: Easier to experiment with problem variants

### Research Enablement

1. **Faster experimentation**: Quicker iteration on model architectures
2. **Larger problems**: Enable research on 20+ user scenarios
3. **Real-time potential**: Move toward real-time network optimization

### Risk 1: Numerical Differences

**Issue**: CVXPY/CuCLARABEL may produce slightly different solutions than CVXOPT

**Mitigation**:
- Establish tolerance thresholds (rtol=1e-3 for power, 1e-4 for objectives)
- Run extensive validation on test datasets
- Compare against ground truth (exhaustive search for n ≤ 5)

### Risk 2: GPU Availability

**Issue**: Not all users have CUDA-compatible GPUs

**Mitigation**:
- Maintain CVXOPT as fallback
- Auto-detect GPU availability
- Provide clear documentation for GPU setup
- Support CPU-only CLARABEL backend

### Risk 3: Reformulation Errors

**Issue**: Incorrect cone constraint formulation may produce wrong results

**Mitigation**:
- Implement comprehensive test suite
- Use CVXPY's Disciplined Convex Programming (DCP) analysis
- Validate against CVXOPT results on diverse problem instances
- Gradual rollout with A/B testing

### Quantitative Metrics

1. **Performance**: ≥2x speedup in training time (measured on CUDA-enabled DevPod)
2. **Accuracy**: Solutions within 0.1% of CVXOPT results
3. **Stability**: Pass all existing unit tests and validation benchmarks
4. **Scalability**: Handle n=20 users without memory overflow

### Qualitative Metrics

1. **Code quality**: Reduced lines of code in optimization functions
2. **Developer experience**: Easier to understand and modify
3. **Community feedback**: Positive reception from contributors

## Implementation Timeline

**Estimated Duration**: 3-4 weeks

- **Week 1**: Setup, Phase 1 + Phase 2.1 (α ∈ [0,1) case)
  - Days 1-2: Environment setup, dependency installation
  - Days 3-5: Implement CVXPY reformulation for simpler case
  - Days 6-7: Initial testing and debugging

- **Week 2**: Phase 2.2 (α = 1 case) + Phase 3.1
  - Days 8-10: Implement complex exponential cone formulation
  - Days 11-12: Integration and backward compatibility
  - Days 13-14: Unit test development

- **Week 3**: Phase 3.2 (Testing) + Phase 4 (Optimization)
  - Days 15-17: Comprehensive testing and validation
  - Days 18-19: Performance optimization and tuning
  - Days 20-21: Batch optimization experiments

- **Week 4**: Phase 5 (Documentation) + Buffer
  - Days 22-24: Documentation updates
  - Days 25-26: Final validation and benchmarking
  - Days 27-28: Buffer for unexpected issues

### Alternative 1: Direct CuCLARABEL Integration

**Pros**: Maximum performance, direct control
**Cons**: Requires manual conic reformulation, higher implementation complexity
**Decision**: Rejected in favor of CVXPY's automatic reformulation

### Alternative 2: JAX + CVXPY

**Pros**: Automatic differentiation, seamless PyTorch integration
**Cons**: Requires learning JAX ecosystem, potential PyTorch interop issues
**Decision**: Consider for future research but CVXPY is more established

### Alternative 3: PyTorch + DiffCP

**Pros**: Native PyTorch integration, differentiable optimization
**Cons**: Limited solver options, less mature ecosystem
**Decision**: Monitor for future consideration

### Alternative 4: CasADi with GPU Plugins

**Pros**: Fast nonlinear optimization, automatic differentiation
**Cons**: Steeper learning curve, less Python-native
**Decision**: Good for future exploration but CVXPY is more accessible

### Technical Documentation

- **CVXPY**: https://www.cvxpy.org/
- **CLARABEL**: https://github.com/oxfordcontrol/Clarabel.rs
- **CuCLARABEL**: https://github.com/cvxgrp/CuClarabel
- **Exponential Cone Programming**: https://docs.mosek.com/modeling-cookbook/expo.html

### Research Papers

- Domahidi et al. (2013): "ECOS: An SOCP solver for embedded systems"
- Goulart et al. (2022): "CLARABEL: An interior point conic solver in Rust"
- Grant & Boyd (2014): "CVX: Matlab Software for Disciplined Convex Programming"

### Related Work

- **NOMA Optimization**: Sun et al. (2019) - "Joint Beamforming and Power Allocation in Downlink NOMA Systems"
- **GPU Acceleration**: Mattingley & Boyd (2012) - "CVXGEN: A code generator for embedded convex optimization"

## Open Questions

1. **Batch Optimization**: Can we reformulate to solve multiple instances simultaneously?
2. **Memory Management**: What's the optimal problem size before GPU memory becomes limiting?
3. **Solver Parameters**: What tolerances and settings optimize for our specific problem structure?
4. **Precision**: Is single-precision (float32) sufficient, or do we need double-precision (float64)?

## Call for Feedback

This is a significant refactoring that will impact the core optimization pipeline. Community feedback is welcome on:

- Technical approach and reformulation strategy
- Testing methodology and success criteria
- Migration timeline and backward compatibility
- Alternative solver options to consider

## Getting Started

**For Contributors**:
1. Review this proposal and provide feedback
2. Test current CVXOPT performance on your hardware (see benchmarking script below)
3. Suggest improvements or concerns
4. Volunteer for specific implementation phases

**Benchmarking Script** (`scripts/benchmark_optimization.py`):
```python
import time
import numpy as np
from resource_allocation_optimization import generate_topology, get_optimal_p

def benchmark_solver(n_users=10, n_trials=100):
    """Benchmark current CVXOPT solver."""
    users = generate_topology(user_number=n_users)
    alpha = 0.5
    noise = 3.981e-15
    
    times = []
    for _ in range(n_trials):
        start = time.time()
        p = get_optimal_p(users, alpha, noise)
        times.append(time.time() - start)
    
    print(f"Users: {n_users}")
    print(f"Mean time: {np.mean(times)*1000:.2f} ms")
    print(f"Std time: {np.std(times)*1000:.2f} ms")
    print(f"Min time: {np.min(times)*1000:.2f} ms")
    print(f"Max time: {np.max(times)*1000:.2f} ms")

if __name__ == "__main__":
    for n in [5, 8, 10, 15, 20]:
        benchmark_solver(n_users=n, n_trials=50)
        print()
```

Run with:
```bash
python scripts/benchmark_optimization.py
```

---

**Issue Labels**: `enhancement`, `performance`, `optimization`, `GPU`, `refactor`

**Priority**: Medium-High

**Effort**: Large (3-4 weeks)

**Impact**: High (core optimization pipeline)
